### ISSUE TITLE
Renderdoc Qt Port without UI or PROJ files

### ISSUE THREAD

**Issue by MattGuerrette** [baldurk/renderdoc#265]
Hello. I wanted to first thank you for creating Renderdoc as I and my colleagues have been using it extensively in our Vulkan development. 

For my day job, I currently work extensively with Qt and would be willing to help develop the Qt port in my spare time, the issue I see however is that you are currently using .ui and .proj files which are generally Qt Creator/qmake specific and turn off developers who would rather work directly in C++ and be able to avoid having to use Qt Creator.

Is there any chance you would be willing to allow development of RenderdocQt without using .ui or .proj files? If so, it would make it a lot easier to develop on both Windows and Linux without needing to use Qt Creator.

Thanks

**Comment by baldurk** [comment]
The thing is the .ui files are the only way to get it working in Qt Creator, and if you hand-create the UI in C++ then there's no WYSIWYG editing. It creates both a barrier to entry for new people wanting to make small changes, and makes it more long-winded/tedious for e.g. me to make changes, doubly so since I'm inexperienced with Qt so any work directly in C++ means constantly ping-ponging to the reference documentation.

So yes, I'd prefer to keep the .ui files and use them for any new windows/dialogs that are created. I'm not sure that the .proj is much of a big deal - you can just generate some build files once (when/if the Qt UI becomes the main UI, I will be committing generated vcproj's to git so that people don't have to install Qt just to build), or if you want you can use whatever alternate build system you would otherwise use if the .proj wasn't there and just keep it local.

**Comment by MattGuerrette** [comment]
Well as far as I know, all the meaningful settings contained in the .proj file are able to be generated using CMake, but since you are using Qt Creator it would be wise to keep a .proj file. I agree that for those not familiar with Qt or used to writing widgets/dialogs by hand it could be tedious. 

Btw, CMake has support for generating Qt ready Visual Studio projects and is what I currently use for Qt development.

**Comment by baldurk** [comment]
Yeh that's the main problem. While it makes things faster for those who are experts in Qt, the tradeoff is against lesser skilled coders and from my experience producing UI code really fast isn't the bottleneck - it tends to be the actual UX design, and the underlying feature implementation. Also since I expect that I'll still be doing the majority of the work on the Qt UI I want to optimise for myself, and that's definitely a preference for WYSIWYG editing :grinning:.

I do not plan to ever use CMake for windows on the project. The plan is that when/if the Qt UI replaces the .NET UI I'll commit the Qt libs/headers from whichever Qt version I use and likewise commit the Qt generated files including visual studio projects, moc files, etc. That way people only need to have Qt installed if they're actually modifying the UI, and otherwise they can just build as normal.

**Comment by MattGuerrette** [comment]
Why do you plan to not ever use CMake for windows? You utilize CMake as a build dependency already and seeing how its one of if not the most popular open source build tool, any developer worth their weight will most likely already have CMake or at the very least, be capable enough to install it. Pushing custom built solution files generally winds up causing a lot of issues long term if they are not consistently updated, and honestly sound like more work than just having cmake generate them for you....

**Comment by baldurk** [comment]
CMake is currently used just on linux, but that only happened recently primarily to ease implementing android support. It is explicitly disallowed on windows.

CMake is much less ubiquitous on windows than you suggest, and even though any developer would know how to install it, that doesn't mean that it's desirable for the project to require that. At the moment there are no other requirements to build than visual studio - you can just clone the repo and build the solution. I really value that simplicity and ease of use, and I don't want to add the extra setup step of cmake. Plus just from personal preference I really dislike cmake based projects on windows.

Committing generated files isn't necessarily a bad thing, and there's no reason why it should be avoided at all costs. In this case it would just be the qt compiled output cpp files, the vcxproj I believe could be manually created the same as any of the other projects that currently exist. As long as some amount of care is taken I don't foresee any problems, but we'll see what happens when the time comes.

**Comment by MattGuerrette** [comment]
Yes, but cmake will  a visual studio project capable of auto-moccing qt source. That is not currently available by default with visual studio without adding a build customization. While those are available for download online, and there are some good tutorials for it, I find it rather cumbersome to either require it or add those build customization files to a repository when the developers of cmake have written that functionality for you. As far as I know, for developing Qt on Windows cmake is definitely becoming the standard.

**Comment by baldurk** [comment]
I don't have any wish though for a visual studio project that can invoke any of the Qt steps (like moc, uic, etc), so cmake being able to do that is immaterial :smile:. What I have in mind is that the visual studio project will not perform any custom build steps (as you point out these are quite clumsy), it will just compile the files that are already submitted to the project.

Anyone who actually needs to modify the UI will need to install Qt - there's no way around that - but that will be a minority of people and they will just use Qt Creator as normal, and commit the updated generated files. Then everyone else just uses a regular visual studio solution with no need for extra steps.

**Comment by MattGuerrette** [comment]
Fair enough. I will just add one more note. Generally, committing generated sources to version control is not only taboo, but doesn't necessarily make much sense. 
Dont believe me? Listen to others sound off: http://stackoverflow.com/questions/893913/should-i-store-generated-code-in-source-control

Furthermore, you will need the Qt dependency to build qrenderdoc regardless if you are working on the UI and moc'ng the source is part of the process. 
I understand that what you are trying to do seems like it would make development easier, but it won't.

**Comment by Honeybunch** [comment]
Matt and I have been discussing this a bit recently. I think it's unfair to say:

> CMake is currently used just on linux, but that only happened recently primarily to ease implementing android support. It is explicitly disallowed on windows.

CMake is a beyond valuable tool on Windows. There are plenty of Linux leaning developers trapped in the Windows world (especially game programmers) who are using GNU-like tools in combination with CMake for build solutions.

I agree that there is value in allowing for development with QT Creator and Visual Studio right out of the box by providing .sln and .proj files. However when dealing with large complicated visual studio projects and trying to also keep a robust Makefile for Linux support can be a pain. Adding a QT project on top of that, that's three different build systems that are all doing very similar things. CMake allows you to generate robust, cross platform build environments at the cost of an extra step to build the project. I think that's a fair price to pay especially since it helps cut generated code from the moc step out of the equation which, as Matt pointed out, is taboo.

**Comment by moradin** [comment]
Having some recent experience with CMake generation of VS project files I have to say I totally agree with Baldur on this. In an ideal world CMake would generate awesome vcxprojs and everyone would be happy. Unfortunately the reality is not this as of today. I talked with few of those who use CMake in gamedev (not a representative amount) and they had similar experiences. So either there is a group of very unlucky people or it's actually not as good as it sound first

**Comment by Honeybunch** [comment]
@moradin I will wholeheartedly admit that CMake isn't perfect but I'm very curious in what ways CMake is lacking for you. I've been using CMake extensively for the past few years with very few complaints other than the sometime vague documentation. The actual solutions and makefiles it has produced have been very solid. [Hatchit](https://github.com/thirddegree/Hatchit) and almost all of its dependencies are leveraging CMake to great effect.

**Comment by baldurk** [comment]
> Fair enough. I will just add one more note. Generally, committing generated sources to version control is not only taboo, but doesn't necessarily make much sense. 
> Dont believe me? Listen to others sound off: http://stackoverflow.com/questions/893913/should-i-store-generated-code-in-source-control

Certainly committing generated sources to VCS is something to be weighed up, but just writing it off as never the right thing to do by appealing to authority is not very useful. If you're doing it for no good reason, then sure that's probably a bad idea,  however there are tradeoffs to consider.

Look for example at [gl_hookset_defs.h](https://github.com/baldurk/renderdoc/blob/master/renderdoc/driver/gl/gl_hookset_defs.h) - this is generated by a small perl script I wrote. Should I require that everyone have ActivePerl installed and run this as a pre-build step? No, because it very rarely changes and it's easy for me (or perhaps in a rare case someone else) to have perl, run it, and commit the updated file. As opposed to requiring every developer to have perl working and in their path, which is a significant inconvenience.

> Furthermore, you will need the Qt dependency to build qrenderdoc regardless if you are working on the UI and moc'ng the source is part of the process. 
> I understand that what you are trying to do seems like it would make development easier, but it won't.

Unless I'm mistaken once the files are generated all that's needed is the standard libraries and headers just like any other SDK, which I will commit to VCS (yes - shock horror, committing binary files to git. There are dozens of stackoverflow links for that one I imagine) so that the visual studio solution continues to be a one-click build. I'm still not seeing how this will make development harder.

> Matt and I have been discussing this a bit recently. I think it's unfair to say:
> > CMake is currently used just on linux, but that only happened recently primarily to ease implementing android support. It is explicitly disallowed on windows.

> CMake is a beyond valuable tool on Windows. There are plenty of Linux leaning developers trapped in the Windows world (especially game programmers) who are using GNU-like tools in combination with CMake for build solutions. :smile:

I don't deny that CMake is a valuable tool for some. What I said was a statement of fact though so I don't see how it can be unfair - renderdoc only uses CMake just on linux with [error messages for use on windows](https://github.com/baldurk/renderdoc/blob/master/CMakeLists.txt#L3-L5), and the changes came in recently - #215 .

> However when dealing with large complicated visual studio projects, trying to also keep a robust Makefile for Linux support can be a pain. Adding a QT project on top of that, that's three different build systems that are all doing very similar things.

Having three different build systems is something I wanted to avoid, which is why the CMakeLists just invokes qmake for qrenderdoc. Yes there will be two separate build systems maintained, but I consider the benefits of this worth the small cost. Adding files is not a complex process and doesn't happen often, and changing build settings is often platform specific anyway. Optimising for making adding files or changing build settings as easy as possible is not the tradeoff I want to make.

Basically as @moradin says I'm not alone in not being a fan of a CMake-based build process, and I also don't believe I am alone in highly prizing the ability to simply clone the source and build immediately, without then finding a need for a third party program to go and install and configure. Given the priority of maintaining this, I am willing to pay what I consider to be reasonable and small costs in other areas. I'm happy to discuss the benefits and drawbacks of this particular approach and particularly see if there are any strictly better ways to improve it, but it's not very productive simply to discount the concerns I and others have with CMake and the goal being aimed for.

**Comment by Honeybunch** [comment]
@baldurk I misread your comment on "windows support being disallowed" as being more of a philosophic argument for some reason. I'm still going to stand by the fact that having multiple build systems is less than optimal, let alone ideal especially since the goal is long-term maintenance., I guess we'll just agree to disagree on that front.

**Comment by MattGuerrette** [comment]
This will be my final comment.

> Unless I'm mistaken once the files are generated all that's needed is the standard libraries and headers just like any other SDK, which I will commit to VCS (yes - shock horror, committing binary files to git. There are dozens of stackoverflow links for that one I imagine) so that the visual studio solution continues to be a one-click build.

Shock? Horror? Yes, I think almost every open source developer would cringe at this statement. You shouldn't just copy open source libraries into your own repo when you could easily submodule them so the project can continue to receive upstream commits. 

Also, why are there Windows SDK files in this repository? Those files are often generated when the SDK is built? Are you planning on manually updating every single one of those files each time the SDK changes?

You discuss making these decisions for maintainability, but this type of project management usually just kills open source projects.

**Comment by cgmb** [comment]
> I see however is that you are currently using .ui and .proj files which are generally Qt Creator/qmake specific and turn off developers who would rather work directly in C++ and be able to avoid having to use Qt Creator.

The XML format for UI files is pretty simple. I actually have found it's sometimes easier to edit them by hand than to use designer. You don't necessarily need to use Qt Creator, though, you probably want to open and close them once in it before committing, just in case.

> commit the updated generated files

I saw the suggestion of committing moc files [mentioned in the wiki](https://github.com/baldurk/renderdoc/wiki/Cross-Platform-UI) a while back. I figured it was a discussion for later, but while we're talking about this, it seems like a bad idea. Each generated file includes something like this near the top:

```
#elif Q_MOC_OUTPUT_REVISION != 67
#error "This file was generated using the moc from 5.5.0. It"
#error "cannot be used with the include files from this version of Qt."
#error "(The moc has changed too much.)"
#endif
```

The Qt project is very careful about binary compatibility. Each release of Qt 5, from Qt 5.0 back in 2012 to Qt 5.6 in 2016 has been binary compatible. However, the `Q_MOC_OUTPUT_REVISION` changes much more frequently, and committing the moc files breaks forward-compatibility.

**Comment by salamanderrake** [comment]
You don't need Qt-Creator to deal with .pro files, all you need is qmake.

**Comment by MattGuerrette** [comment]
.pro files are not really the issue. Its the .ui files. However, i would prefer to avoid using .pro files in any Qt project, as Qt's cmake support is quite good and QtCreator is pretty trash.
